File: src/com/morph/engine/util/Timer.java
package com.morph.engine.util;

/**
 * Created by Fernando on 2/13/2017.
 */
public class Timer {
    private float interval;
    private float accumulator;
    private Runnable action;
    private boolean running;

    public Timer(float interval, Runnable action) {
        this.interval = interval;
        this.action = action;
        this.running = false;
    }

    public void step(float dt) {
        if (accumulator >= interval && running) {
            action.run();
            accumulator = 0;
        }

        if (running)
            accumulator += dt;
    }

    public void setInterval(float interval) {
        this.interval = interval;
    }

    public void setAction(Runnable action) {
        this.action = action;
    }

    public void start() {
        running = true;
    }

    public void stop() {
        running = false;
    }
}

File: src/main/kotlin/com/morph/engine/util/Timer.kt
package com.morph.engine.util

/**
 * Created by Fernando on 2/13/2017.
 */
class Timer(private var interval: Float, private var action: Runnable) {
    private var accumulator: Float = 0.toFloat()
    private var running: Boolean = false

    init {
        this.running = false
    }

    fun step(dt: Float) {
        if (accumulator >= interval && running) {
            action.run()
            accumulator = 0f
        }

        if (running)
            accumulator += dt
    }

    fun setInterval(interval: Float) {
        this.interval = interval
    }

    fun setAction(action: Runnable) {
        this.action = action
    }

    fun start() {
        running = true
    }

    fun stop() {
        running = false
    }

    fun restart() {
        running = true
        accumulator = 0f
    }
}
